File: src/services/StravaService.js
/**
 * Service pour l'intégration avec Strava
 */

// Configuration Strava
const STRAVA_CONFIG = {
  clientId: '159311',
  clientSecret: '6e2acab2c5731c20a946fe672afdfc8a6b6485f1',
  redirectUri: window.location.origin,
  scope: 'read',
};

// Clé de stockage local pour les statistiques Strava
const STRAVA_STATS_KEY = 'project_fat_loss_strava_stats';

// Informations d'authentification de l'utilisateur
let stravaAuth = {
  accessToken: localStorage.getItem('strava_access_token') || '09dae37d7b963c3fe7fd1478e6c0315260e45404',
  refreshToken: localStorage.getItem('strava_refresh_token') || 'fb1159763f37433a9a135f70495360e87fa8201a',
  expiresAt: localStorage.getItem('strava_expires_at') || '2025-05-13T04:02:30Z'
};

/**
 * Initialise le service Strava
 */
export function initStravaService() {
  // Vérifier si le token est présent dans le localStorage
  if (!stravaAuth.accessToken) {
    console.log('Aucun token Strava trouvé');
    return false;
  }

  // Vérifier si le token est expiré
  const expiresAt = new Date(stravaAuth.expiresAt).getTime();
  const now = new Date().getTime();
  
  if (now >= expiresAt) {
    console.log('Token Strava expiré, rafraîchissement nécessaire');
    return refreshToken();
  }
  
  console.log('Service Strava initialisé avec succès');
  return true;
}

/**
 * Rafraîchit le token d'accès Strava lorsqu'il expire
 * @returns {Promise<boolean>} - true si le token a été rafraîchi avec succès
 */
async function refreshToken() {
  try {
    const response = await fetch('https://www.strava.com/oauth/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json'
      },
      body: JSON.stringify({
        client_id: STRAVA_CONFIG.clientId,
        client_secret: STRAVA_CONFIG.clientSecret,
        refresh_token: stravaAuth.refreshToken,
        grant_type: 'refresh_token'
      })
    });

    if (!response.ok) {
      throw new Error(`Erreur HTTP: ${response.status}`);
    }

    const data = await response.json();
    
    // Mettre à jour les informations d'authentification
    stravaAuth = {
      accessToken: data.access_token,
      refreshToken: data.refresh_token,
      expiresAt: new Date(Date.now() + data.expires_in * 1000).toISOString()
    };

    // Sauvegarder dans localStorage
    localStorage.setItem('strava_access_token', data.access_token);
    localStorage.setItem('strava_refresh_token', data.refresh_token);
    localStorage.setItem('strava_expires_at', stravaAuth.expiresAt);
    
    console.log('Token Strava rafraîchi avec succès');
    return true;
  } catch (error) {
    console.error('Erreur lors du rafraîchissement du token Strava:', error);
    return false;
  }
}

/**
 * Authentifie l'utilisateur avec Strava via OAuth
 * @returns {Promise<boolean>} - true si l'authentification a réussi
 */
export function authenticateWithStrava() {
  // Si déjà authentifié et token valide, ne rien faire
  if (stravaAuth.accessToken) {
    const expiresAt = new Date(stravaAuth.expiresAt).getTime();
    if (expiresAt > Date.now()) {
      return Promise.resolve(true);
    }
  }
  
  // Rediriger vers la page d'authentification Strava
  const authUrl = `https://www.strava.com/oauth/authorize?client_id=${STRAVA_CONFIG.clientId}&redirect_uri=${encodeURIComponent(STRAVA_CONFIG.redirectUri)}&response_type=code&scope=${STRAVA_CONFIG.scope}`;
  
  // Ouvrir dans une nouvelle fenêtre
  const authWindow = window.open(authUrl, 'StravaAuth', 'width=600,height=800');
  
  return new Promise((resolve) => {
    // Vérifier périodiquement si l'utilisateur a terminé l'authentification
    const checkAuth = setInterval(() => {
      if (authWindow.closed) {
        clearInterval(checkAuth);
        
        // Vérifier si l'authentification a réussi (le token est présent)
        if (localStorage.getItem('strava_access_token')) {
          stravaAuth.accessToken = localStorage.getItem('strava_access_token');
          stravaAuth.refreshToken = localStorage.getItem('strava_refresh_token');
          stravaAuth.expiresAt = localStorage.getItem('strava_expires_at');
          resolve(true);
        } else {
          resolve(false);
        }
      }
    }, 500);
  });
}

/**
 * Enregistre un entraînement sur Strava
 * @param {Object} workout - Les données de l'entraînement
 * @returns {Promise<Object>} - Les données de l'entraînement enregistré
 */
export async function syncWorkout(workout) {
  // Vérifier si l'utilisateur est authentifié
  if (!stravaAuth.accessToken) {
    const authenticated = await authenticateWithStrava();
    if (!authenticated) {
      throw new Error("L'utilisateur n'est pas authentifié avec Strava");
    }
  }

  // Vérifier si le token est expiré
  const expiresAt = new Date(stravaAuth.expiresAt).getTime();
  if (Date.now() >= expiresAt) {
    const refreshed = await refreshToken();
    if (!refreshed) {
      throw new Error("Impossible de rafraîchir le token Strava");
    }
  }

  try {
    // Formatter les données pour Strava
    const stravaActivity = {
      name: workout.name || "Entraînement Project Fat Loss",
      type: "WeightTraining",
      start_date_local: workout.date ? new Date(workout.date).toISOString() : new Date().toISOString(),
      elapsed_time: workout.duration || 1800, // Par défaut 30 minutes
      description: workout.desc || `Entraînement de musculation: ${workout.exerciseCount || 0} exercices, ${workout.calories || 0} calories brûlées`,
    };

    // Envoyer l'entraînement à Strava
    const response = await fetch('https://www.strava.com/api/v3/activities', {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${stravaAuth.accessToken}`,
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(stravaActivity)
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(`Erreur Strava: ${errorData.message || response.statusText}`);
    }

    const data = await response.json();
    console.log('Entraînement synchronisé avec Strava:', data);
    
    return data;
  } catch (error) {
    console.error('Erreur lors de la synchronisation avec Strava:', error);
    throw error;
  }
}

/**
 * Récupère les entraînements de l'utilisateur depuis Strava
 * @param {number} limit - Nombre d'entraînements à récupérer
 * @returns {Promise<Array>} - Liste des entraînements
 */
export async function getStravaActivities(limit = 10) {
  // Vérifier si l'utilisateur est authentifié
  if (!stravaAuth.accessToken) {
    const authenticated = await authenticateWithStrava();
    if (!authenticated) {
      throw new Error("L'utilisateur n'est pas authentifié avec Strava");
    }
  }

  // Vérifier si le token est expiré
  const expiresAt = new Date(stravaAuth.expiresAt).getTime();
  if (Date.now() >= expiresAt) {
    const refreshed = await refreshToken();
    if (!refreshed) {
      throw new Error("Impossible de rafraîchir le token Strava");
    }
  }

  try {
    const response = await fetch(`https://www.strava.com/api/v3/athlete/activities?per_page=${limit}`, {
      headers: {
        'Authorization': `Bearer ${stravaAuth.accessToken}`
      }
    });

    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(`Erreur Strava: ${errorData.message || response.statusText}`);
    }

    const data = await response.json();
    return data;
  } catch (error) {
    console.error('Erreur lors de la récupération des activités Strava:', error);
    throw error;
  }
}

/**
 * Synchronise l'historique des entraînements avec Strava
 * @param {number} limit - Nombre d'entraînements à synchroniser (du plus récent au plus ancien)
 * @returns {Promise<Object>} - Résultats de la synchronisation
 */
export async function syncWorkoutHistory(limit = 5) {
  try {
    // Vérifier si l'utilisateur est authentifié
    if (!stravaAuth.accessToken) {
      const authenticated = await authenticateWithStrava();
      if (!authenticated) {
        throw new Error("L'utilisateur n'est pas authentifié avec Strava");
      }
    }

    // Vérifier si le token est expiré
    const expiresAt = new Date(stravaAuth.expiresAt).getTime();
    if (Date.now() >= expiresAt) {
      const refreshed = await refreshToken();
      if (!refreshed) {
        throw new Error("Impossible de rafraîchir le token Strava");
      }
    }

    // Récupérer l'historique des entraînements depuis le stockage local
    const { getWorkoutHistory } = await import('./WorkoutStorage');
    const workouts = getWorkoutHistory();
    
    // Prendre uniquement les plus récents
    const recentWorkouts = workouts.sort((a, b) => 
      new Date(b.date).getTime() - new Date(a.date).getTime()
    ).slice(0, limit);
    
    // Synchroniser chaque entraînement
    const results = {
      total: recentWorkouts.length,
      synced: 0,
      failed: 0,
      details: []
    };
    
    for (const workout of recentWorkouts) {
      try {
        // Vérifier si cet entraînement contient déjà un ID Strava (pour éviter les doublons)
        if (workout.stravaId) {
          results.details.push({
            title: workout.title,
            status: 'skipped',
            reason: 'Déjà synchronisé'
          });
          continue;
        }
        
        // Préparer les données pour Strava
        const stravaActivity = {
          name: workout.title || "Entraînement Project Fat Loss",
          type: "WeightTraining",
          start_date_local: workout.date,
          elapsed_time: workout.duration || workout.exerciseCount * 180 || 1800, // 3 min par exercice ou 30min par défaut
          description: `Entraînement de musculation: ${workout.exerciseCount} exercices, ${workout.calories} calories brûlées, ${workout.weightLifted || 0}kg soulevés`,
        };
        
        // Envoyer l'entraînement à Strava
        const response = await fetch('https://www.strava.com/api/v3/activities', {
          method: 'POST',
          headers: {
            'Authorization': `Bearer ${stravaAuth.accessToken}`,
            'Content-Type': 'application/json'
          },
          body: JSON.stringify(stravaActivity)
        });
        
        if (!response.ok) {
          const errorData = await response.json();
          throw new Error(`Erreur Strava: ${errorData.message || response.statusText}`);
        }
        
        const data = await response.json();
        
        // Mettre à jour l'entrainement local avec l'ID Strava
        const { updateWorkoutWithStravaId } = await import('./WorkoutStorage');
        await updateWorkoutWithStravaId(workout.id, data.id);
        
        results.synced++;
        results.details.push({
          title: workout.title,
          status: 'success',
          stravaId: data.id
        });
      } catch (error) {
        console.error(`Erreur lors de la synchronisation de l'entraînement ${workout.title}:`, error);
        results.failed++;
        results.details.push({
          title: workout.title,
          status: 'failed',
          error: error.message
        });
      }
    }
    
    return results;
  } catch (error) {
    console.error('Erreur lors de la synchronisation de l\'historique:', error);
    throw error;
  }
}

/**
 * Vérifie si l'utilisateur est connecté à Strava
 * @returns {boolean} - true si l'utilisateur est connecté
 */
export function isConnectedToStrava() {
  return !!stravaAuth.accessToken;
}

/**
 * Déconnecte l'utilisateur de Strava
 */
export function disconnectFromStrava() {
  localStorage.removeItem('strava_access_token');
  localStorage.removeItem('strava_refresh_token');
  localStorage.removeItem('strava_expires_at');
  localStorage.removeItem(STRAVA_STATS_KEY);
  
  stravaAuth = {
    accessToken: null,
    refreshToken: null,
    expiresAt: null
  };
}

/**
 * Récupère les statistiques d'entraînements synchronisés avec Strava
 * @returns {Promise<Object>} - Les statistiques
 */
export async function getStravaSyncStats() {
  try {
    // Vérifier si les stats sont en cache
    const cachedStats = localStorage.getItem(STRAVA_STATS_KEY);
    if (cachedStats) {
      const stats = JSON.parse(cachedStats);
      const cacheAge = Date.now() - stats.lastUpdate;
      // Utiliser le cache si moins de 1 heure
      if (cacheAge < 3600000) {
        return stats;
      }
    }

    // Si pas connecté, retourner des stats vides
    if (!isConnectedToStrava()) {
      return {
        totalSynced: 0,
        totalDistance: 0,
        totalCalories: 0,
        lastSync: null,
        lastUpdate: Date.now()
      };
    }

    // Chercher les 100 dernières activités
    const activities = await getStravaActivities(100);
    const stats = {
      totalSynced: activities.length,
      totalDistance: activities.reduce((sum, act) => sum + (act.distance || 0), 0),
      totalCalories: activities.reduce((sum, act) => sum + (act.calories || 0), 0),
      lastSync: activities[0]?.start_date || null,
      lastUpdate: Date.now()
    };

    // Mettre en cache
    localStorage.setItem(STRAVA_STATS_KEY, JSON.stringify(stats));
    return stats;
  } catch (error) {
    console.error('Erreur lors de la récupération des statistiques Strava:', error);
    throw error;
  }
}

File: projectfatloss_flutter/lib/shared/services/strava_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'preferences_service.dart';

/// Service d'intégration Strava
/// Reproduit le StravaService.js de la PWA
class StravaService {
  // Configuration OAuth Strava (identique à la PWA)
  static const String _clientId = '159311';
  static const String _clientSecret =
      'VOTRE_CLIENT_SECRET'; // À configurer
  static const String _redirectUri =
      'https://votreapp.com/strava-callback'; // À configurer

  static const String _authUrl = 'https://www.strava.com/oauth/authorize';
  static const String _tokenUrl = 'https://www.strava.com/oauth/token';
  static const String _apiUrl = 'https://www.strava.com/api/v3';

  /// Génère l'URL d'autorisation OAuth
  static String getAuthorizationUrl() {
    final params = {
      'client_id': _clientId,
      'redirect_uri': _redirectUri,
      'response_type': 'code',
      'scope': 'activity:write,activity:read_all',
      'approval_prompt': 'auto',
    };

    final queryString = params.entries
        .map((e) => '${e.key}=${Uri.encodeComponent(e.value)}')
        .join('&');

    return '$_authUrl?$queryString';
  }

  /// Échange le code d'autorisation contre un access token
  static Future<Map<String, dynamic>?> exchangeCodeForToken(
    String code,
  ) async {
    try {
      final response = await http.post(
        Uri.parse(_tokenUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'client_id': _clientId,
          'client_secret': _clientSecret,
          'code': code,
          'grant_type': 'authorization_code',
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        // Sauvegarder les tokens
        await PreferencesService.setStravaAccessToken(data['access_token']);
        await PreferencesService.setStravaRefreshToken(data['refresh_token']);
        await PreferencesService.setStravaTokenExpiry(data['expires_at']);
        await PreferencesService.setStravaConnected(true);

        print('✅ Strava connecté avec succès');
        return data;
      } else {
        print('❌ Erreur échange token: ${response.statusCode}');
        return null;
      }
    } catch (e) {
      print('❌ Erreur échange token: $e');
      return null;
    }
  }

  /// Rafraîchit l'access token
  static Future<bool> refreshAccessToken() async {
    final refreshToken = PreferencesService.stravaRefreshToken;
    if (refreshToken.isEmpty) return false;

    try {
      final response = await http.post(
        Uri.parse(_tokenUrl),
        headers: {'Content-Type': 'application/json'},
        body: jsonEncode({
          'client_id': _clientId,
          'client_secret': _clientSecret,
          'refresh_token': refreshToken,
          'grant_type': 'refresh_token',
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);

        await PreferencesService.setStravaAccessToken(data['access_token']);
        await PreferencesService.setStravaRefreshToken(data['refresh_token']);
        await PreferencesService.setStravaTokenExpiry(data['expires_at']);

        print('✅ Token Strava rafraîchi');
        return true;
      } else {
        print('❌ Erreur refresh token: ${response.statusCode}');
        return false;
      }
    } catch (e) {
      print('❌ Erreur refresh token: $e');
      return false;
    }
  }

  /// Vérifie si le token est expiré et le rafraîchit si nécessaire
  static Future<bool> ensureValidToken() async {
    final expiryTimestamp = PreferencesService.stravaTokenExpiry;
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;

    // Si le token expire dans moins de 5 minutes, le rafraîchir
    if (expiryTimestamp - now < 300) {
      return await refreshAccessToken();
    }

    return true;
  }

  /// Envoie un entraînement vers Strava
  static Future<bool> uploadWorkout({
    required String name,
    required String description,
    required DateTime startDate,
    required int durationSeconds,
    required String activityType,
    double? calories,
  }) async {
    if (!PreferencesService.stravaConnected) {
      print('⚠️ Strava non connecté');
      return false;
    }

    try {
      // Vérifier et rafraîchir le token si nécessaire
      if (!await ensureValidToken()) {
        print('❌ Token invalide');
        return false;
      }

      final accessToken = PreferencesService.stravaAccessToken;

      final body = {
        'name': name,
        'type': activityType, // 'WeightTraining'
        'start_date_local': startDate.toIso8601String(),
        'elapsed_time': durationSeconds,
        'description': description,
        'trainer': 1,
        'commute': 0,
      };

      final response = await http.post(
        Uri.parse('$_apiUrl/activities'),
        headers: {
          'Authorization': 'Bearer $accessToken',
          'Content-Type': 'application/json',
        },
        body: jsonEncode(body),
      );

      if (response.statusCode == 201) {
        final data = jsonDecode(response.body);
        print('✅ Activité Strava créée: ${data['id']}');
        return true;
      } else {
        print('❌ Erreur upload Strava: ${response.statusCode}');
        print('Body: ${response.body}');
        return false;
      }
    } catch (e) {
      print('❌ Erreur upload Strava: $e');
      return false;
    }
  }

  /// Récupère les activités Strava
  static Future<List<Map<String, dynamic>>> getActivities({
    int page = 1,
    int perPage = 30,
  }) async {
    if (!PreferencesService.stravaConnected) {
      return [];
    }

    try {
      if (!await ensureValidToken()) {
        return [];
      }

      final accessToken = PreferencesService.stravaAccessToken;

      final response = await http.get(
        Uri.parse('$_apiUrl/athlete/activities?page=$page&per_page=$perPage'),
        headers: {
          'Authorization': 'Bearer $accessToken',
        },
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.cast<Map<String, dynamic>>();
      } else {
        print('❌ Erreur récupération activités: ${response.statusCode}');
        return [];
      }
    } catch (e) {
      print('❌ Erreur récupération activités: $e');
      return [];
    }
  }

  /// Récupère les infos de l'athlète
  static Future<Map<String, dynamic>?> getAthleteInfo() async {
    if (!PreferencesService.stravaConnected) {
      return null;
    }

    try {
      if (!await ensureValidToken()) {
        return null;
      }

      final accessToken = PreferencesService.stravaAccessToken;

      final response = await http.get(
        Uri.parse('$_apiUrl/athlete'),
        headers: {
          'Authorization': 'Bearer $accessToken',
        },
      );

      if (response.statusCode == 200) {
        return jsonDecode(response.body);
      } else {
        print('❌ Erreur récupération athlete: ${response.statusCode}');
        return null;
      }
    } catch (e) {
      print('❌ Erreur récupération athlete: $e');
      return null;
    }
  }

  /// Déconnecte Strava
  static Future<void> disconnect() async {
    try {
      // Révoquer le token
      final accessToken = PreferencesService.stravaAccessToken;

      if (accessToken.isNotEmpty) {
        await http.post(
          Uri.parse('https://www.strava.com/oauth/deauthorize'),
          headers: {
            'Authorization': 'Bearer $accessToken',
          },
        );
      }

      // Nettoyer les préférences
      await PreferencesService.setStravaConnected(false);
      await PreferencesService.setStravaAccessToken('');
      await PreferencesService.setStravaRefreshToken('');
      await PreferencesService.setStravaTokenExpiry(0);

      print('✅ Strava déconnecté');
    } catch (e) {
      print('❌ Erreur déconnexion Strava: $e');
    }
  }

  /// Synchronise l'historique des entraînements
  static Future<void> syncWorkoutHistory(
    List<Map<String, dynamic>> workouts,
  ) async {
    if (!PreferencesService.stravaConnected) return;

    print('🔄 Synchronisation de ${workouts.length} entraînements...');

    int successCount = 0;
    int failCount = 0;

    for (final workout in workouts) {
      try {
        final success = await uploadWorkout(
          name: workout['dayTitle'] ?? 'Entraînement',
          description:
              'Calories: ${workout['totalCalories']} kcal\nPoids total: ${workout['totalWeight']} kg',
          startDate: DateTime.parse(workout['date']),
          durationSeconds: workout['duration'] ?? 0,
          activityType: 'WeightTraining',
          calories: (workout['totalCalories'] as num?)?.toDouble(),
        );

        if (success) {
          successCount++;
        } else {
          failCount++;
        }

        // Pause entre les uploads pour éviter le rate limiting
        await Future.delayed(const Duration(seconds: 1));
      } catch (e) {
        print('❌ Erreur sync workout: $e');
        failCount++;
      }
    }

    print(
      '✅ Sync terminée: $successCount réussis, $failCount échoués',
    );
  }

  /// Vérifie si Strava est connecté et valide
  static Future<bool> isConnectedAndValid() async {
    if (!PreferencesService.stravaConnected) return false;
    return await ensureValidToken();
  }
}